mlir/lib/Quantum/Transforms/wrapper_for_wrapper.cpp

```diff
@@ -0,0 +1,220 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mlir/Conversion/LLVMCommon/TypeConverter.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/Support/FormatVariadic.h"
+
+#include "Quantum/Transforms/Passes.h"
+#include "Quantum/Transforms/Patterns.h"
+
+using namespace mlir;
+using namespace catalyst::quantum;
+
+namespace catalyst {
+namespace quantum {
+
+static size_t countCallsites(LLVM::LLVMFuncOp op)
+{
+    size_t count = 0;
+    op.walk([&](LLVM::CallOp callOp) { count++; });
+    return count;
+}
+
+static bool callsiteHasAttribute(LLVM::LLVMFuncOp op)
+{
+    bool retval = false;
+    ModuleOp moduleOp = cast<ModuleOp>(op->getParentOp());
+    if (!moduleOp)
+        return retval;
+    op.walk([&](LLVM::CallOp callOp) {
+        auto callee = callOp.getCallee();
+        if (callee) {
+            auto stringRef = callee.value().data();
+            Operation *fnDecl = moduleOp.lookupSymbol<LLVM::LLVMFuncOp>(stringRef);
+            retval = (bool)(fnDecl->getAttrOfType<UnitAttr>(
+                LLVM::LLVMDialect::getEmitCWrapperAttrName()));
+        }
+    });
+    return retval;
+}
+
+struct WrapperForWrapperTransform : public OpRewritePattern<LLVM::LLVMFuncOp> {
+    using OpRewritePattern<LLVM::LLVMFuncOp>::OpRewritePattern;
+
+    LogicalResult match(LLVM::LLVMFuncOp op) const override;
+    void rewrite(LLVM::LLVMFuncOp op, PatternRewriter &rewriter) const override;
+};
+
+LogicalResult WrapperForWrapperTransform::match(LLVM::LLVMFuncOp op) const
+{
+    std::string _mlir_ciface = "_mlir_ciface_";
+    size_t _mlir_ciface_len = _mlir_ciface.length();
+    auto symName = op.getSymName();
+    const char *symNameStr = symName.data();
+    size_t symNameLength = strlen(symNameStr);
+    // Filter based on name.
+    if (symNameLength <= _mlir_ciface_len)
+        return failure();
+
+    bool nameMatches = 0 == strncmp(symNameStr, _mlir_ciface.c_str(), _mlir_ciface_len);
+    if (!nameMatches)
+        return failure();
+
+    // Function must only contain a single function call.
+    size_t callsites = countCallsites(op);
+    if (callsites != 1)
+        return failure();
+
+    // The only call must contain the emitCWrapper attribute
+    bool hasAttribute = callsiteHasAttribute(op);
+
+    return hasAttribute ? success() : failure();
+}
+
+static bool functionHasReturns(LLVM::LLVMFuncOp op)
+{
+    bool result = false;
+    op.walk([&](LLVM::CallOp callOp) { result = (bool)callOp.getResult(); });
+    return result;
+}
+
+static bool functionHasInputs(LLVM::LLVMFuncOp op)
+{
+    bool result = false;
+    op.walk([&](LLVM::CallOp callOp) { result = !callOp.getOperandTypes().empty(); });
+    return result;
+}
```

Comment:
All this walking seems a bit strange, especially because of the implicit assumption of a single call in the function.
Couldn't the call reference just be reused once it's found?
Probably not too important though given the size of the function.

Reply:
The function that we are manipulating was generated by the compiler, and it follows the specification from `_mlir_ciface` which guarantees it is only a wrapper function and only contains 1 callsite to the function it wraps.

Reply:
I'll see if I can reuse the symbol attribute.

Reply:
https://github.com/PennyLaneAI/catalyst/pull/106/commits/15f408edf3b516c29fedcd3351e8dfadb06dc281 (This also addresses other comments).

Reply:
Looks nice, but would the use of `SymbolTable::getSymbolUses` that I linked above not simplify the search for the callee?

Reply:
I'll investigate but to me this seems like the wrong meaning of "uses". We are not looking for the uses of `_mlir_ciface`. We are looking for whatever `_mlir_ciface` calls. It is not a use of `_mlir_ciface`.

Reply:
(I was mixing it up with the earlier version that iterated through the module to find a call for the given FuncOp)